### PR TITLE
Changed the behaviour of boundary checking to be compatible with other AbstractArray types.

### PR DESCRIPTION
**Pull request by RainerHeintzmann**
In `ShiftedArray` and `CircShiftedArray` the  boundaries are now checked according to the new array size. This allows them to be used for `PaddedArrays` and more. However, it is a breaking change. It may be possible to think about two variants, one with the old way of boundary checking and one with the standard way.

### CODE REVIEW

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:50]: This seems wrong for ranges with offsets (eg `2:10`). Maybe best to restrict to `AbstractUnitRange` and use `length(range)`.

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:48]:
```suggestion
```

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:54]: Probably better to leave as is, or just do `map(bringwithin, idxs, ranges)` and remove the method below. I think `map` is a bit more idiomatic for `Tuple`s than `broadcast`.

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:75]:
```suggestion
```

File: test/runtests.jl
Comment (piever) [inline test/runtests.jl:38]: Maybe add a test for axes with offsets, like `ShiftedArrays.bringwithin(11, 2:7)`

File: test/runtests.jl
Comment (piever) [inline test/runtests.jl:68]:
```suggestion
    @test !checkbounds(Bool, sv, 123)  # RH changed this behaviour
```

**Comment by piever** [comment]
Thank you, I've added a few comment. I think it'd be relevant to do some benchmarking to see if with the new behavior one can reduce the overhead for `CircShiftedArrays`. I think it's worth to compare

```julia
v, w = rand(1000), rand(1000)
w .= v .- ShiftedArrays.circshift(v, 1)
```

with

```julia
for i in eachindex(w)
    w[i] = i == 1 ? v[1] - v[end] : v[i] - v[i-1]
end
```

to see what is the overhead due to `mod`. If it is too much, it may be worth to think if it can be made more efficient by a simple `if` condition (and doing the `mod` computation upon construction).

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:104]:
```suggestion
```

**Review by piever — CHANGES_REQUESTED** [review]

File: src/circshiftedarray.jl
Comment (roflmaostc) [reply src/circshiftedarray.jl]: Hey,
I checked timings:
```julia
julia> using BenchmarkTools

julia> f(w, v) = begin
       for i in eachindex(w)
                  w[i] = i == 1 ? v[1] - v[end] : v[i] - v[i-1]
              end
           end
f (generic function with 1 method)

julia> g(w, v) = w .= v .- ShiftedArrays.circshift(v, 1);

julia> v, w = rand(1000), rand(1000);

julia> @btime g($w, $v); # is based on the modulo idea 
  1.425 μs (0 allocations: 0 bytes)

julia> @btime f($w, $v);
  652.564 ns (0 allocations: 0 bytes)
  ```
  
I then switched the implementation to a non-mod one where `shifts` is processed at the beginning:
```
julia> @btime g($w, $v);
  457.609 ns (0 allocations: 0 bytes)
```
Performance is therefore ~3x better.
Unfortunately `shifts` is not equal to the input anymore. Is that a problem? Semantically it is still the same.

**Comment by roflmaostc** [comment]
Hey,
I checked timings:
```julia
julia> using BenchmarkTools

julia> f(w, v) = begin
       for i in eachindex(w)
                  w[i] = i == 1 ? v[1] - v[end] : v[i] - v[i-1]
              end
           end
f (generic function with 1 method)

julia> g(w, v) = w .= v .- ShiftedArrays.circshift(v, 1);

julia> v, w = rand(1000), rand(1000);

julia> @btime g($w, $v); # is based on the modulo idea 
  1.425 μs (0 allocations: 0 bytes)

julia> @btime f($w, $v);
  652.564 ns (0 allocations: 0 bytes)
  ```
  
I then switched the implementation to a non-mod one where `shifts` is processed at the beginning:
```
julia> @btime g($w, $v);
  457.609 ns (0 allocations: 0 bytes)
```
Performance is therefore ~3x better.
Unfortunately `shifts` is not equal to the input anymore. Is that a problem? Semantically it is still the same.

**Comment by piever** [comment]
> I then switched the implementation to a non-mod one where shifts is processed at the beginning. Performance is therefore ~3x better.

That's great! As a micro-optimization (we should really optimize `getindex` as much as possible, as it is basically the only performance critical code in the package), it'd be good to also try a branch-free version. For example, once we know that `0 <= shifts < size(parent)`, I would recommend benchmarking something like

```julia
function newindex(old_idx::Int, sz::Int, shift::Int)
    diff = sz - shift # I imagine the compiler will figure out that this is always the same computation throughout the for loop
    return ifelse(old_idx > diff, old_idx - diff, old_idx + shift)
end
```

or some related trick multiplying by a boolean, eg `bringwithin(i, sz) = i - (i > sz) * sz`, instead of the ternary operator `?`. The reason is that, even though it looks like the ternary operator may do less operations, it has to "branch", which would end up preventing some optimizations on the for loop, so it would probably generate slower code in the end ([this blog posts](https://tamaspapp.eu/post/branch_prediction/) explains it much more clearly.)

> Unfortunately `shifts `is not equal to the input anymore. Is that a problem? Semantically it is still the same.

That's why I'd like to have that change here: since we are breaking one thing, we should break both at the same time and bump the version number. I think it's perfectly fine as a behavior, esp. if we document that `shifts` is the original shift modulo the size of the array.

**Comment by roflmaostc** [comment]
I'm not sure if `newindex` works for OffsetAxis? I believe you need to know start and end position of your axis.

Nevertheless, I tried `bringwithin` using `ifelse` but it didn't provide a great speedup. The reason is probably, that the branch prediction fails only once in the index position where the switch between left and rights occurs. For all other indices the branch is always the same and branch prediction should get that correct.
Does that make sense?

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:40]:
```suggestion
        n = map(mod, _padded_tuple(p, n), size(p))
```

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:33]: This comment should probably be part of the docstring of `CircShiftedArray`.

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:38]: Same as above (possibly unifying in a unique line in the docstring).

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:61]:
```suggestion
    return ifelse(ind_with_offset < first(ranges), ind_with_offset + length(ranges), ind_with_offset)
```
(no need for `broadcast` or branching). Here it should actually be `range`, because it is only one.

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:69]:
```suggestion
    ind = offset(shifts(s), x)
    i = map(bringwithin, ind, axes(s))
```

File: src/circshiftedarray.jl
Comment (piever) [inline src/circshiftedarray.jl:77]:
```suggestion
    ind = offset(shifts(s), x)
    i = map(bringwithin, ind, axes(s))
```

File: test/runtests.jl
Comment (piever) [inline test/runtests.jl:65]: Here and below the comment should be removed before merging

File: test/runtests.jl
Comment (piever) [inline test/runtests.jl:76]:
```suggestion
    @test shifts(sv) == (2, 0)
```

**Review by piever — CHANGES_REQUESTED** [review]

**Comment by piever** [comment]
Thanks for the changes! I've added a few comments. Please try to respect the original formatting in terms of whitespace inside tuples and newlines (single newline within test blocks or method definitions).

**Comment by roflmaostc** [comment]
Another thing, we could incorporate:
Wouldn't it be nice being able to remove the circshifted wrapper if we shift backwards the same shift?
Something like this:
```julia
if iszero.(shifts)
     return parent(csa)
end
```

In `fftshift` and `ifftshift` it would be quite nice, so that simply the circshifted wrapper is removed because `ifft` only works with a `ShiftedArray` if you do an allocate upfront calling it. So removing the wrapper once the shift is 0, prevents that.

File: src/circshiftedarray.jl
Comment (piever) [reply src/circshiftedarray.jl:34]: I guess the comments (here and below) can probably be removed now that this information is in the docstring.

**Comment by piever** [comment]
Looks good to me, I'll just wait a little bit to merge in case somebody has objections.

> So the constructor should dispatch on the `CircShiftedArray` and add the shifts instead of wrapping them. And if 0, return the plain array.

I feel a bit uncomfortable changing the behavior of the constructor, but the various syntactic sugar functions (ie, `lag`, `lead`, `circshift`) could definitely be a bit smarter and just add the shifts if the underlying vector is a `(Circ)ShiftedArray`. Unwrapping if `shift` is zero is probably not ideal, because it would not be type stable.

**Comment by roflmaostc** [comment]
I agree on the type issue, could be problematic.

We are discussing for [FourierTools.jl](https://github.com/bionanoimaging/FourierTools.jl/) whether we introduce something like a `FFTShift`and a `IFFTShift` type that could be removed, added, etc. 
We probably start implementing it there, and if it works fine, we could also add that to ShiftedArrays (if wanted, it also makes sense to have that only in FourierTools.jl)

File: test/runtests.jl
Comment (piever) [inline test/runtests.jl:64]:
```suggestion
    sv[3] = 12
```
(trailing whitespace)

**Comment by roflmaostc** [comment]
Ups, forced push the wrong repo, but restored it from a local version :smile:

**Comment by roflmaostc** [comment]
However, what I was trying to do:
In the case of concatenation of ShiftedArrays (like `fftshift` and `ifftshift`do in combination) we could instead of generate a nested `CircShiftedArray` simply update the shift, right?
The change would be minor, we just need a new constructor dispatching on `CircShiftedArray`.
That would prevent the type instability because the `CircShiftedArray` would be still here, but just with zero shift (which can be easily removed in our FourierTools.jl then.

Since this would be breaking as well, what do you think of incorporating?

**Review by piever — APPROVED** [review]

**Comment by piever** [comment]
LGTM, if you are happy with it I can go ahead and merge it. We should probably discuss the nested `lag`, `lead` and `circshift` in a separate PR / issue. (About that, I would definitely prefer to keep the `CircShiftedArray` constructor as is, but we could consider special-casing `circshifted(c::CircshiftedArray, shift)`, `lag(s::ShiftedArray, shift)`, and `lead(s::ShiftedArray, shift)` .)

**Comment by roflmaostc** [comment]
I believe we can merge, yeah.
This special-casing sounds pretty nice imo.

**Pull request merged**
